homeassistant/components/legrandinone.py

```diff
@@ -0,0 +1,480 @@
+    """Look at event to determine type of device.
+
+    Async friendly.
+    """
+    if event.get('type') == EVENT_KEY_COMMAND:
+        return EVENT_KEY_COMMAND
+
+    return 'unknown'
+
+
+async def async_setup(hass, config):
+    """Set up the IOBL component."""
+    from iobl.protocol import create_iobl_connection
+    import serial
+
+    # Allow entities to register themselves by device_id to be looked up when
+    # new IOBL events arrive to be handled
+    hass.data[DATA_ENTITY_LOOKUP] = {
+        EVENT_KEY_COMMAND: defaultdict(list),
+    }
+
+    # Allow platform to specify function to register new unknown devices
+    hass.data[DATA_DEVICE_REGISTER] = {}
+    hass.data[DATA_DEVICE_REGISTER][EVENT_KEY_COMMAND] = {}
+
+    async def async_send_command(call):
+        """Send legrandinone command."""
+        _LOGGER.debug('LegrandInOne command for %s', str(call.data))
+
+        data = {
+            'legrand_id': call.data.get(CONF_DEVICE_ID),
+            'who': call.data.get(CONF_DEVICE_TYPE),
+            'mode': call.data.get(CONF_COMM_MODE),
+            'media': call.data.get(CONF_COMM_MEDIA),
+            'unit': call.data.get(CONF_DEVICE_UNIT),
+            'what': call.data.get(CONF_COMMAND)
+        }
+
+        data['type'] = call.data.get(CONF_PACKET_TYPE)
+        if data['type'] is None:
+            # Default to bus_command packet if not specified.
+            data['type'] = 'bus_command'
+
+        data['values'] = call.data.get(CONF_DIMENSION_VALUES)
+
+        if not await LegrandInOneCommand.send_command(data):
+            _LOGGER.error('Failed LegrandInOne command for %s', str(call.data))
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_SEND_COMMAND, async_send_command,
+        schema=SEND_COMMAND_SCHEMA)
+
+    @callback
+    def event_callback(event):
+        """Handle incoming legrandinone events.
+
+        IOBL events arrive as dictionaries of varying content
+        depending on their type. Identify the events and distribute
+        accordingly.
+        """
+        event_type = event.get('type')
+        _LOGGER.debug('event of type %s: %s', event_type, event)
+
+        # Don't propagate non entity events (eg: version string, ack response)
+        if event_type not in hass.data[DATA_ENTITY_LOOKUP]:
+            _LOGGER.debug('unhandled event of type: %s', event_type)
+            return
+
+        # Lookup entities who registered this device id as device id or alias
+        event_id = event.get('legrand_id', None)
+
+        entity_ids = hass.data[DATA_ENTITY_LOOKUP][event_type][event_id]
+
+        if entity_ids:
+            # Propagate event to every entity matching the device id
+            for entity in entity_ids:
+                _LOGGER.debug('passing event to %s', entity)
+                async_dispatcher_send(hass,
+                                      SIGNAL_HANDLE_EVENT.format(entity),
+                                      event)
+        else:
```

Comment:
If we `return` at the end of the above if block, we can remove the `else:` here and outdent the following code.

Reply:
I'm not sure to understand the aim of doing this ? Wouldn't this make the code less readable ?

Reply:
No, flat is better than nested.